File: TDD_practice/js/bank.js
const Money = require('./money');

class Bank {

    constructor() {
        this.exchangeRates = new Map();
    }
    addExchangeRate(from, to, rate) {
        this.exchangeRates.set(`${from}->${to}`, rate);
    }

    convert(money, to) {
        if (money.currency === to) {
            return new Money(money.amount, money.currency);
        }

        let key = `${money.currency}->${to}`;
        let rate = this.exchangeRates.get(key);

        if (rate === undefined) {
            throw new Error(key);
        }
        return new Money(money.amount * rate, to);
    }
}

module.exports = Bank;

File: tdd-project/py/bank.py
from money import Money


class Bank:
    def __init__(self):
        self.exchange_rates = {}

    def add_exchange_rate(self, from_currency, to_currency, rate):
        self.exchange_rates[f'{from_currency}->{to_currency}'] = rate

    def convert(self, money, to_currency):
        key = f'{money.currency}->{to_currency}'
        if money.currency == to_currency:
            return Money(money.amount, to_currency)

        if key in self.exchange_rates:
            return Money(money.amount * self.exchange_rates[key], to_currency)

        raise KeyError(key)
